[Cheng Zi Wei Keith] iP

KeithCZW opened the pull request:
# Welcome to Duke!
> "To eat an apple and expect to have a cake after is foolishness" - Sun Tzu, probably

Duke is a chatbot that will help you manage tasks! It's,
- text-based
- easy to learn
- ~~FAST~~ __GIGA FAST__ to use!

How to use it? All you have to do is,
1. download it from [here](https://github.com/KeithCZW/ip/releases/tag/A-Jar).
2. double-click it.
3. add your tasks
4. tasks = managed 😄

It's also **FREE!!!!**

Features include:
- [x] Managing tasks
- [ ] Managing deadlines (coming soon)
- [ ] Reminders (coming soon)

---

If you're a Java programmer, you can use it to practice Java too. Here's the `main` method:

```java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

KeithCZW committed:
Level-1 commit, greet, echo, exit

KeithCZW committed:
Level 2 commit, Add,List

KeithCZW committed:
Level 3 Commit, Mark as done

KeithCZW committed:
Level 4 Commit, ToDo, Event, Deadline

KeithCZW committed:
Automated Text UI Testing

KeithCZW committed:
Level-5 Error Handling

KeithCZW committed:
Level-6 Delete

KeithCZW committed:
Level-7 Save

KeithCZW committed:
Level-8 Dates and Times

KeithCZW committed:
Merge branch 'branch-Level-8'

KeithCZW committed:
Parsing date from console now has error checks

KeithCZW committed:
Add time parsing to Deadline and Event classes

KeithCZW committed:
Add OOP for tasks

KeithCZW committed:
Split into packages

KeithCZW committed:
Remove some imports

KeithCZW committed:
Add JUnit Test for Task

KeithCZW committed:
Allow for JAR file

KeithCZW committed:
Add Javadocs to all classes

KeithCZW committed:
Coding style updates

KeithCZW committed:
Fix some javadocs spacing

KeithCZW committed:
Add Find functionality

KeithCZW committed:
Merge branch 'branch-A-CodingStandard'

# Conflicts:
#	src/main/java/duke/Storage.java

KeithCZW committed:
Merge branch 'branch-Level-9'

# Conflicts:
#	src/main/java/duke/Parser.java

KeithCZW committed:
Add Javadoc for Find function

KeithCZW committed:
Merge remote-tracking branch 'origin/add-gradle-support'

KeithCZW committed:
Add Gradle support

KeithCZW committed:
Add Checkstyle support

KeithCZW committed:
Add javaFx GUI to Duke

KeithCZW committed:
Removed some comments

KeithCZW committed:
Move javaFx code from Duke to DialogContainer

iztanpy commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,103 @@
+import duke.task.Task;
+import duke.task.Todo;
+
+import java.io.*;
```
Might not be accurate to import java generic, maybe you could import the specific classes utilised instead?

KeithCZW replied:
Oops! I missed that. Will fix it.

iztanpy commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,151 @@
+    public String parseTodo(String input) throws DukeException {
+        String[] split = input.split(" ");
+        if (split.length < 2) {
+            throw new DukeException("OH NO! The description of event cannot be empty.");
```
I did something similar but an alternative could be

throw new DukeException(e.getMessage());

iztanpy reviewed (commented):
Nicely done! Just suggested some small alternatives and coding standard fixes 👍

lamwj98 commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,302 @@
+                        throw new DukeException("No such task exists.");
+                    }
+                    storage.OverwriteFile(tasks.getTasks());
+                    break;
+                case TODO: {
+                    Todo newTodo = new Todo(parser.parseTodo(input));
+                    tasks.addTask(newTodo);
+                    ui.sayMessage(
+                            "Got it. I've added this task:\n" + newTodo + "\n" + "Now you have " + tasks.getTasks()
+                                    .size() + " tasks in the list.");
+                    storage.OverwriteFile(tasks.getTasks());
+                    break;
+                }
+                case DEADLINE:
+                    String[] deadlineInput = parser.parseDeadline(input);
+                    Deadline newDeadline = new Deadline(deadlineInput[0], deadlineInput[1]);
+                    tasks.addTask(newDeadline);
+                    ui.sayMessage(
+                            "Got it. I've added this task:\n" + newDeadline + "\n" + "Now you have " + tasks.getTasks()
+                                    .size() + " tasks in the list.");
+                    storage.OverwriteFile(tasks.getTasks());
+                    break;
+                case EVENT:
+                    String[] eventInput = parser.parseEvent(input);
+                    Event newEvent = new Event(eventInput[0], eventInput[1]);
+                    tasks.addTask(newEvent);
+                    ui.sayMessage(
+                            "Got it. I've added this task:\n" + newEvent + "\n" + "Now you have " + tasks.getTasks()
+                                    .size() + " tasks in the list.");
+                    storage.OverwriteFile(tasks.getTasks());
+                    break;
+                case FIND:
+                    String word = parser.parseFind(input);
+                    ui.listTasks(tasks.findTasks(word));
+                    break;
+                case ERROR:
+                    throw new DukeException("Sorry :( I don't know what this means.");
+                }
+            } catch (DukeException e) {
+                ui.sayMessage(e.getErrorMsg());
```
would it be better to handle this cases somewhere else or create another class to avoid populating the duke class?

KeithCZW replied:
Yeah I will probably encapsulate the functionality in another class soon.

lamwj98 commented on src/main/java/duke/DukeException.java:
```diff
@@ -0,0 +1,26 @@
+     *
+     * @return The error message
+     */
+    public String getErrorMsg() {
+        return errorMsg;
+    }
```
what is the purpose of creating a getErrorMsg function when there is a .getMessage() function from exception?

lamwj98 commented on src/main/java/duke/task/Deadline.java:
```diff
@@ -0,0 +1,94 @@
+     *
+     * @return time/date
+     */
+    public String getBy() {
```
naming of variable is not easily understand, maybe getDue would be a better name ?

KeithCZW replied:
Yeah perhaps some of the naming is too vague, I will change that in the next push.

lamwj98 commented on src/main/java/duke/task/Event.java:
```diff
@@ -0,0 +1,94 @@
+ * Event task that inherits from Task
+ */
+public class Event extends Task {
+    protected LocalDate at;
```
same issue, naming of variable not easily understandable, would `date` be a better name?

lamwj98 commented on src/main/java/duke/task/Event.java:
```diff
@@ -0,0 +1,94 @@
+     *
+     * @return time/date
+     */
+    public String getAt() {
```
same as above, would getEventDate or getDate be better?

lamwj98 commented on src/main/java/duke/task/Task.java:
```diff
@@ -0,0 +1,84 @@
+    }
+
+    /**
+     * Overloaded toString function
```
Should be overwrite instead of overload

lamwj98 commented on src/main/java/duke/task/Task.java:
```diff
@@ -0,0 +1,84 @@
+        return "[" + getStatusIcon() + "] " + description;
+    }
+
+    /**
+     * Empty getFileString function for child classes
+     *
+     * @return Nothing
+     */
+    public String getFileString() {
+        return "";
+    }
+}
```
would it be better to declare as an abstract method?

KeithCZW replied:
Yes, fair enough. I will change this!

lamwj98 reviewed (commented):
Overall code is easy to read and understandable. Some improvements needed on variables namings to make methods more understandable.

KeithCZW committed:
Allow build.gradle to compile jar files.

KeithCZW committed:
Added assert functionality to all Parser functions.
This ensures that input passed into the parser is always correct.

Removed unnecessary import statements from all files.
This ensures that only the required libraries are being imported.
Removing all the extra import statements will help keep the file
clean.

Removed unnecessary else statements from Parser
This will help to prevent arrowhead code blocks from forming

KeithCZW committed:
Merge pull request #1 from KeithCZW/branch-A-Assertions

Added assert functionality to all Parser functions.

KeithCZW committed:
Remove more unnecessary imports

KeithCZW committed:
Deadlines and Events now can be snoozed

KeithCZW committed:
Remove unused comments

KeithCZW committed:
Even more OOP
Encapsulate the commands into classes

KeithCZW committed:
Use JavaFX to make UI images

KeithCZW committed:
Add screenshot for ui

KeithCZW committed:
Set theme jekyll-theme-midnight

KeithCZW committed:
Image Ui.png changed

KeithCZW committed:
Merge branch 'master' of https://github.com/KeithCZW/ip

KeithCZW committed:
Remove UI.png

KeithCZW committed:
Add Ui.png

KeithCZW committed:
Add User guide

KeithCZW committed:
Update docs readme
